### PR TITLE
[ASTPrinter] Fix printing of nested typealias types and make it consistent with printing of nominal types.

### PR DESCRIPTION
**Pull request by akyrtzi**
#### What's in this pull request?

```
This fixes several issues:
- By default parent types of alias types are not printed which results in
    - Erroneous fixits, for example when casting to 'Notification.Name' from a string, which ends up adding erroneous cast
      as "Name(rawValue: ...)"
    - Hard to understand types in code-completion results and diagnostics
- When printing with 'fully-qualified' option typealias types are printed erroneously like this "<PARENT>.Type.<TYPEALIAS>"

The change make typealias printing same as nominal types and addresses the above.
```
#### Resolved bug number: ([SR-](https://bugs.swift.org/browse/SR-))

rdar://27809198

---

Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>

Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please smoke test |
| All supported platforms | @swift-ci Please smoke test and merge |
| OS X platform | @swift-ci Please smoke test OS X platform |
| Linux platform | @swift-ci Please smoke test Linux platform |

A smoke test on macOS does the following:
1. Builds the compiler incrementally.
2. Builds the standard library only for macOS. Simulator standard libraries and
   device standard libraries are not built.
3. lldb is not built.
4. The test and validation-test targets are run only for macOS. The optimized
   version of these tests are not run.

A smoke test on Linux does the following:
1. Builds the compiler incrementally.
2. Builds the standard library incrementally.
3. lldb is built incrementally.
4. The swift test and validation-test targets are run. The optimized version of these
   tests are not run.
5. lldb is tested.

**Validation Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please test |
| All supported platforms | @swift-ci Please test and merge |
| OS X platform | @swift-ci Please test OS X platform |
| OS X platform | @swift-ci Please benchmark |
| Linux platform | @swift-ci Please test Linux platform |

**Lint Testing**

| Language | Comment |
| --- | --- |
| Python | @swift-ci Please Python lint |

Note: Only members of the Apple organization can trigger swift-ci.
</details>

…stent with printing of nominal types.

This fixes several issues:
- By default parent types of alias types are not printed which results in
  - Erroneous fixits, for example when casting to 'Notification.Name' from a string, which ends up adding erroneous cast
    as "Name(rawValue: ...)"
  - Hard to understand types in code-completion results and diagnostics
- When printing with 'fully-qualified' option typealias types are printed erroneously like this "<PARENT>.Type.<TYPEALIAS>"

The change make typealias printing same as nominal types and addresses the above.

### CODE REVIEW

**Comment by akyrtzi** [comment]
@swift-ci smoke test OS X platform

**Comment by akyrtzi** [comment]
@swift-ci test OS X platform

File: lib/AST/ASTPrinter.cpp [hunk 1]
```diff
@@ -3590,12 +3590,24 @@ class TypePrinter : public TypeVisitor<TypePrinter> {
...
+    auto ParentNominal = ParentDC ?
+      ParentDC->getAsNominalTypeOrNominalTypeExtensionContext() : nullptr;
+
+    if (ParentNominal) {
+      Type ParentTy = ParentNominal->getType();
+      if (auto *MT = ParentTy->getAs<AnyMetatypeType>()) {
+        // Don't print as '<PARENT_TYPENAME>.Type'.
+        // FIXME: There should be a convenient API to get a TypeDecl as a type
+        // without the meta type.
+        ParentTy = Type(MT->getInstanceType());
```
Comment (benlangmuir) [inline lib/AST/ASTPrinter.cpp]: `getDeclaredType()` will give you the non-meta type.

**Comment by benlangmuir** [comment]
LGTM

**Comment by akyrtzi** [comment]
@swift-ci smoke test OS X platform

**Comment by akyrtzi** [comment]
@swift-ci Please smoke test OS X

**Comment by akyrtzi** [comment]
@swift-ci smoke test OS X platform

**Comment by akyrtzi** [comment]
@swift-ci Please test

**Pull request merged**